### PR TITLE
Qeubo

### PR DESCRIPTION
**Pull request by flokde**

### CODE REVIEW

File: pbmohpo/utils.py [hunk 1]
```diff
@@ -152,3 +152,39 @@ def color_generator():
...
+        by `q` `d`-dimensional decision vectors.
+
+    new_y: torch.Tensor
+        An n x 1 tensor of training outputs. Each of the `n` responses is an
+        integer between 0 and `q-1` indicating the decision vector selected by
+        the user.
+    """
+    helper_list_X = []
+
+    for duel in y:
```
Comment (ja-thomas) [inline pbmohpo/utils.py:183]: List comprehension?

Comment (ja-thomas) [reply pbmohpo/utils.py:183]: helper_list_X = [torch.stack([X[int(duel[0])], X[int(duel[1])]]) for duel in y]

File: pbmohpo/utils.py [hunk 2]
```diff
@@ -152,3 +152,39 @@ def color_generator():
 
     for color in colors:
         yield color
+
+
+def convert_torch_archive_for_variational_preferential_gp(
+    X: torch.Tensor, y: torch.Tensor
+):
+    """
+    Get bounds of hyperparameters in the format botorch needs
```
Comment (ja-thomas) [inline pbmohpo/utils.py:161]: copy/paste from other helper function?

File: pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from typing import Optional
```
Comment (ja-thomas) [inline pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py:1]: I would move all stuff from likelihoods, models and acqusition functions to optimizers (maybe in one file botorch_utils or smth like that). 

This feels like a lot of folders where only a single class is present

File: pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+from botorch.sampling import MCSampler
+from botorch.utils.transforms import (
+    concatenate_pending_points,
+    t_batch_mode_transform,
+)
+from torch import Tensor
+
+
+class qExpectedUtilityOfBestOption(MCAcquisitionFunction):
+    r"""Expected Utility of Best Option (qEUBO).
```
Comment (ja-thomas) [inline pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py:18]: r?

File: pbmohpo/optimizers/eubo.py [hunk 5]
```diff
@@ -128,3 +135,90 @@ def _surrogate_proposal(self, archive: Archive, n: int) -> List[CS.Configuration
...
+        Size of the initial design, if not specified, two times the number of HPs is used
+    """
+
+    def __init__(
+        self,
+        config_space: CS.ConfigurationSpace,
+        initial_design_size: Optional[int] = None,
+    ) -> None:
+
+        # TODO: Is this really needed or can I pass None along?
```
Comment (ja-thomas) [inline pbmohpo/optimizers/eubo.py:165]: I don't think you need that

File: pbmohpo/optimizers/eubo.py [hunk 6]
```diff
@@ -128,3 +135,90 @@ def _surrogate_proposal(self, archive: Archive, n: int) -> List[CS.Configuration
...
+        -------
+        CS.Configuration:
+            Proposed Configuration
+
+        """
+
+        X, _ = archive.to_torch()
+        y = torch.Tensor(archive.comparisons)
+
+        X, y = convert_torch_archive_for_variational_preferential_gp(X, y)
```
Comment (ja-thomas) [inline pbmohpo/optimizers/eubo.py:193]: If we don't need this in any other class either move the code in the _surrogate_proposal or make it a member of function of qEUBO

File: pbmohpo/optimizers/eubo.py [hunk 7]
```diff
@@ -128,3 +135,90 @@ def _surrogate_proposal(self, archive: Archive, n: int) -> List[CS.Configuration
...
+
+        model = VariationalPreferentialGP(X, y)
+        model.train()
+        model.likelihood.train()
+
+        mll = VariationalELBO(
+            likelihood=model.likelihood,
+            model=model,
+            num_data=2 * model.num_data,
```
Comment (ja-thomas) [inline pbmohpo/optimizers/eubo.py:202]: why 2 * model.num_data?

File: pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py (hunk 4 above)
Comment (flokde) [reply pbmohpo/acquisition_functions/qexpected_utility_of_best_option.py:18]: Von den facebook Leuten geklaut (die machen das scheinbar standardmäßig). Laut doku braucht man raw nur, wenn man backslashes im docstring hat. Also kann es eigentlich auch rausnehmen.

File: pbmohpo/optimizers/eubo.py (hunk 7 above)
Comment (flokde) [reply pbmohpo/optimizers/eubo.py:201]: Taken from facebook implementation. I figured we start with their recommended values. But honestly don't know what the best value is here (I owuld hope it is somewhat robust). Happy to discuss though.

**Comment by sumny** [comment]
- [ ] add tests
- [ ] add docs

**Pull request merged**